### PR TITLE
expose lr scheduler python bindings for on device training.

### PR DESCRIPTION
**Pull request by AdamLouly**
### Description
Exposing LR Scheduler python bindings for on device training.

### CODE REVIEW

File: orttraining/orttraining/python/training/api/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Licensed under the MIT License.
+# lr_scheduler.py
+
+from onnxruntime.capi import _pybind_state as C
+
+
+class LRScheduler:
+    """
+    Class for running Learning Scheduler Step in Training.
+    This class is a wrapper of LRScheduler Class.
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/lr_scheduler.py:11]: Let's provide more details here. What kind of LR scheduler is this (linear, ...)? How should the user use it?

Also, let's avoid reference to the C++ class by saying that this is a wrapper of the underlying C++ class. Python users may not know the C++ class.

For example:

```py
"""Linearly updates the learning rate in the optimizer

    The linear learning rate scheduler decays the learning rate by linearly updated
    multiplicative factor from the initial learning rate set on the training session to 0. The decay
  is performed after the initial warm up phase where the learning rate is linearly incremented
  from to the initial learning rate provided.

    Args:
        optimizer (:obj:`training_api.Optimizer`): User's onnxruntime training Optimizer
        warmup_step_count (int): The number of steps in the warm up phase.
        total_step_count (int): The total number of training steps.

```

File: orttraining/orttraining/python/training/api/lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    def __init__(self, optimizer, warmup_step_count, total_step_count) -> None:
+        """
+        Initializes LRScheduler with the optimizer model and other parameters.
+        """
+        self._scheduler = C.LRScheduler(optimizer._optimizer, warmup_step_count, total_step_count)
+
+    def step(self):
+        """
+        Run LRScheduler Step.
+        """
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/lr_scheduler.py:23]: These comments are not very useful. Let's provide more context on what happens when `step()` is invoked.

File: orttraining/orttraining/python/training/api/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+class LRScheduler:
+    """
+    Class for running Learning Scheduler Step in Training.
+    This class is a wrapper of LRScheduler Class.
+    """
+
+    def __init__(self, optimizer, warmup_step_count, total_step_count) -> None:
+        """
+        Initializes LRScheduler with the optimizer model and other parameters.
+        """
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/lr_scheduler.py:17]: Let's remove this documentation since the class documentation should cover most of the intent.

File: orttraining/orttraining/python/training/api/lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+from onnxruntime.capi import _pybind_state as C
+
+
+class LRScheduler:
+    """
+    Class for running Learning Scheduler Step in Training.
+    This class is a wrapper of LRScheduler Class.
+    """
+
+    def __init__(self, optimizer, warmup_step_count, total_step_count) -> None:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/lr_scheduler.py:14]: Do we not need to set the initial learning rate? Let's keep the behavior consistent with the C training apis.

File: orttraining/orttraining/python/training/api/lr_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# lr_scheduler.py
+
+from onnxruntime.capi import _pybind_state as C
+
+
+class LRScheduler:
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/api/lr_scheduler.py:8]: Should this be `LinearLRScheduler`?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 6]
```diff
@@ -162,6 +162,35 @@ struct TrainingConfigurationResult {
...
+  std::shared_ptr<onnxruntime::training::api::Optimizer> GetOptimizer() { return optimizer_; }
+
+  virtual ~PyOptimizer() {}
+
+ protected:
+  PyOptimizer(std::shared_ptr<onnxruntime::training::api::Optimizer> optimizer) {
+    optimizer_ = std::move(optimizer);
+  }
+
+  std::shared_ptr<onnxruntime::training::api::Optimizer> optimizer_;
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:190]: Why do we need these to be protected?

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 7]
```diff
@@ -162,6 +162,35 @@ struct TrainingConfigurationResult {
...
+        GetTrainingORTEnv(), provider);
+  }
+
+  Status Step() const { return optimizer_->Step(); }
+  Status SetLearningRate(float lr) { return optimizer_->SetLearningRate(lr); }
+  float GetLearningRate() { return optimizer_->GetLearningRate(); }
+
+  std::shared_ptr<onnxruntime::training::api::Optimizer> GetOptimizer() { return optimizer_; }
+
+  virtual ~PyOptimizer() {}
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:183]: Can we avoid these by making the optimizer public. The pybind functions can just reference the optimizer and call the approriate functions.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 8]
```diff
@@ -162,6 +162,35 @@ struct TrainingConfigurationResult {
   optional<std::string> loss_scale_input_name;
 };
 
+#ifdef ENABLE_TRAINING_ON_DEVICE
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:165]: Please refer to PR #13964 which might change the use of this macro. Whichever PR goes in later might need to update.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 6 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:190]: should It be private?

Comment (baijumeswani) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:190]: it can be public (in the case of struct, you can avoid providing any specifier). Just access the variable directly from outside as required.

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 8 above)
Comment (AdamLouly) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:165]: do we need to change this after the other PR gets merged or before,
so it should just be in ENABLE_TRAINING flag from what I understand from the PR.

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 9]
```diff
@@ -914,29 +928,44 @@ void addObjectMethodsForTraining(py::module& m, ExecutionProviderRegistrationFn
...
+  lr_scheduler.def(py::init([](PyOptimizer* optimizer,
+                               int64_t total_step_count,
+                               int64_t warmup_step_count,
+                               float initial_lr) {
+                ORT_THROW_IF_ERROR(optimizer->optimizer_->SetInitialLearningRate(initial_lr));
+
+                auto scheduler = std::make_unique<onnxruntime::training::api::LinearLRScheduler>(
+                    optimizer->optimizer_, warmup_step_count, total_step_count);
+
+                return scheduler;
```
Comment (baijumeswani) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:964]: optional:

```suggestion
                return std::make_unique<onnxruntime::training::api::LinearLRScheduler>(
                    optimizer->optimizer_, warmup_step_count, total_step_count);
```

**Review by baijumeswani — APPROVED** [review]

File: orttraining/orttraining/python/orttraining_pybind_state.cc (hunk 8 above)
Comment (askhade) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:165]: No, it should be ENABLE_TRAINING_APIS...

**Review by baijumeswani — APPROVED** [review]
